captcha refactor complete (#15)
CorneliaKelinske/first_blog: posts/2022/08-25-adapter_pattern.md

==title==

> Same, same, but different - using Behaviour for writing reusable and adaptable code 

Reviewer: ```suggestion
Same same, but different - using Behaviours to write extensible code
```

https://theculturetrip.com/asia/thailand/articles/same-same-but-different-the-origins-of-thailands-tourist-catchphrase/

==author==
Cornelia Kelinske

==description==

> I recently worked on an assignment where I had to periodically fetch data from different APIs and store it in a unified format in a database. In order to avoid duplication and to only write one GenServer module for triggering the API calls and database operations, I used Behaviour, aka the Adapter Pattern.

Reviewer: ```suggestion
I recently worked on an assignment where I had to periodically fetch data from different APIs and store it in a unified format in a database. To avoid duplication and only write one GenServer module for triggering the API calls and database operations, I used Behaviours, aka the Adapter Pattern.
```

==tags==
coding, elixir, otp, Behaviour

==body==

# 1. The task at hand

> We need to write some code that allows us to periodically query a number of different APIs to obtain event data and persist it in a database. I live in Canada, so let's say the events are hockey games (DISCLAIMER: I know nothing about hockey!). We want to get information on the home team, the away team and the start time. To keep things simple, we will only query 2 APIs for this example. They both return the information we want; however, API A is returning it in this format:

Reviewer: ```suggestion
We need to write some code that allows us to periodically query several different APIs to obtain event data and then persist the responses in a database. I live in Canada, so let's say the events are hockey games (DISCLAIMER: I know nothing about hockey!). We want to get information on the home team, the away team, and the start time. To keep things simple, we will only query 2 APIs for this example. They both return the information we want; however, API A is returning it in this format:
```

```
[ { "home_team": "Edmonton Oilers", "away_team": "Vancouver Canucks", "start_at": "2022-11-19T09:00:00Z", "created_at": "2018-12-19T09:00:00Z"} ]
```

while API B is returning this:

```

> [ { "teams": "Edmonton Oilers" - "Vancouver Canucks", "start_at": "2022-11-19T09:00:00Z", "created_at": "2018-12-19T09:00:00Z" } ]

Reviewer: ```suggestion
[ { "teams": "Edmonton Oilers - Vancouver Canucks", "start_at": "2022-11-19T09:00:00Z", "created_at": "2018-12-19T09:00:00Z" } ]
```

```
In addition, API B allows us to pass in a `last_checked_at` search parameter, which gives us the option to only fetch match data created after the time we last checked, thus avoiding the retrieval of large amounts of duplicate data.

# 2. The puzzle pieces

Before I start writing any code, I like to think about which pieces I will need to put in place to complete the task.

> In the case of our hockey games, we will need a supervised process that ensures that the APIs are queried periodically, code for making the API calls, code for standardizing the return data and last but not least the database components, namely a table and a schema for the match data.

Reviewer: ```suggestion
In the case of our hockey games, we will need a supervised process that queries the APIs at regular intervals, code for making the API calls, code for standardizing the return data and last but not least, the database components: namely a table and a schema for the match data.
```

Reviewer: Wonder if this should be a bullet point list, as it might be easier to parse for the reader.

# 3. Arranging the pieces

> Now that we know what pieces we need, let's figure out how to put them together. The database side is not different from any other project, so we'll just consider it done. 

Reviewer: ```suggestion
Now that we know which pieces we need let's figure out how to put them together. The database side is no different from any other project, so we'll just consider it done. 
```

> Next up is our process. We could use a task that triggers the API calls, stores the fetched data in the database and then calls itself again after we run `Process.sleep/1` with the required amount n of milliseconds. However, since the API calls and data processing operations also take a small amount of time, we would end up querying the API not every n milliseconds but instead every n milliseconds + the time required to do all the other stuff.

Reviewer: ```suggestion
Next up is our process. We could use a task that triggers the API calls, stores the fetched data in the database and then calls itself again after we run `Process.sleep/1` with the required amount n of milliseconds. However, since the API calls and data processing operations also take a small amount of time, we would end up querying the API not every n milliseconds but instead every n milliseconds + the processing time of the task.
```

> A GenServer is more reliable since we can utilize `Process.send_after/4` to have the GenServer call itself reliably every n milliseconds.

Reviewer: ```suggestion
A GenServer is more predictable as we can utilize `Process.send_after/4` to have the GenServer call itself at regular intervals independent of the time it spent processing.
```

Since the GenServer is doing the same things for both APIs, we can start the same GenServer twice; all we have to do is register the two instances of our GenServer under different names.

We also need code for calling our APIs. I decided to create one API module per queried API. While the two API modules have the same job (calling the API, fetching JSON data, decoding JSON data, standardizing data, returning standardized to the GenServer), they are operating under different circumstances: different HTTP addresses, plus the fact that API B also takes in the additional search parameter. 

> This is the big moment: enter Behaviour.

Reviewer: ```suggestion
This is the big moment: enter Behaviours.
```

I'm pretty sure the concept is plural

https://elixir-lang.org/getting-started/typespecs-and-behaviours.html

> Since we assume that both our API modules will perform the same actions and send data in the same unified format back to the GenServer, we can go one step further and not just assume but require that they do so. 

Reviewer: ```suggestion
Since we assume that both our API modules will perform the same actions and send data in the same unified format back to the GenServer, we can go one step further and not just assume but enforce that they do so at compile time.
```

> To this end, we can use another module to define the desired Behaviour and implement said Behaviour in our API modules. Now, we cannot only be sure that our existing API modules act in a certain way, but we can also easily add further API modules in the future. All we have to do is implement the same Behaviour for them, thus ensuring that we don't accidentally add a rogue API module that acts different from the rest and risks bringing down its associated GenServer.

Reviewer: ```suggestion
To this end, we can use another module to define the desired Behaviour via "callbacks" and implement said Behaviour in our API modules. Now, we cannot only be sure that our existing API modules act in a certain way, but we can also easily add further API modules in the future. All we have to do is implement the same Behaviour for them, thus making it easy to extend current functionality by adding additional modules.
```

But enough with the talk. Let's look at some code!

# 4. The code

**General organization**

> ```

Reviewer: ```suggestion
```bash
```

lib
│   ├── my_app
│   │   ├── application.ex
│   │   ├── config.ex

> │   │   ├── match_data_processors

Reviewer: Are they actually data processors? Or just clients?

Author: I leave them as processors. they query and convert after all

│   │   │   ├── apis
│   │   │   │   ├── client.ex
│   │   │   │   ├── api_a.ex
│   │   │   │   ├── api_b.ex
│   │   │   │   
│   │   │   └── helpers
│   │   │       └── processing_helpers.ex
│   │   ├── matches
│   │   │   └── match.ex
│   │   ├── matches.ex
│   │   ├── match_handler.ex

```

The `match_handler.ex` file is the GenServer module, `match.ex` and `matches.ex` define our the match `Ecto.Schema` and the corresponding context module, and we can also see the `client.ex` (this is where the Behaviour is defined), our two API modules and a `processing_helpers.ex` file, which contains some helper functions accessed by both API modules.

**GenServer**

First stop: `application.ex`, where the GenServer is added to the Supervision tree:

```

> defmodule MyApplication.Application do

Reviewer: ```suggestion
defmodule MyApp.Application do
```

  
  use Application

  @impl true
  def start(_type, _args) do

>     children = [MyApp.Repo] ++ match_handlers()

Reviewer: ```suggestion
    children = [MyApp.Repo | match_handlers()]
```

    
    opts = [strategy: :one_for_one, name: MyApp.Supervisor]
    Supervisor.start_link(children, opts)
  end

  if Mix.env() === :test do
    def match_handlers do
      []
    end
  else

>     @api_modules MyApp.Config.api_modules()

Reviewer: You don't have the `MyApp.Config` module defined in this blog post right?

Author: Right. I will hope it is still clear. (I have the YOLO batch for a reason lol)

    def match_handlers do
      Enum.map(
        @api_modules,
        &Supervisor.child_spec({MyApp.MatchHandler, &1}, id: &1)
      )
    end
  end
end
Footer
```
We are defining our `match_handlers` aka the GenServers for our APIs by `Enum.mapping` over our available API_modules and calling `Supervisor.child_spec/2` on them, where we set the respective API_module as the GenServer module's ID.
We can then add the `match_handlers` to our list of children in the `start/2` function. 

Second stop: the actual GenServer module, `match_handler.ex`:

```
defmodule MyApp.MatchHandler do
  
  use GenServer
  
  alias MyApp.Matches

  # Client

  @spec start_link( module()) :: :ignore | {:error, any} | {:ok, pid}
  def start_link(api_module) do
    GenServer.start_link(__MODULE__, api_module, name: :"match_handler#{inspect(api_module)}")
  end

  # Server

  @impl GenServer
  def init(api_module) do
    Process.send(self(), :match_query, [])
    {:ok, api_module.initial_state()}
  end

  @impl GenServer
  def handle_info(
        :match_query,
        %{api_module: api_module} = state

>       ) do

Reviewer: This could be a single line and more readable
